File: luiz/ba2f.py
#! /usr/bin/env python

import random

from ba2c import profile_most_probable
from ba2d import score
from ba2e import profile_matrix


def randomized_motif_search(dna, k, t):
    motifs = []
    for seq in dna:
        rand_pos = random.randint(0, len(seq) - k)
        motifs.append(seq[rand_pos : rand_pos + k])

    best = motifs
    score_best = score(best)

    while True:
        profile = profile_matrix(motifs, k)
        motifs = [profile_most_probable(seq, k, profile) for seq in dna]

        score_motifs = score(motifs)
        if score_motifs < score_best:
            best = motifs
            score_best = score_motifs
        else:
            return best, score_best


def repeated_randomized_motif_search(dna, k, t, *, times=1000):
    best, score_best = randomized_motif_search(dna, k, t)
    for _ in range(times - 1):
        new_best, new_score = randomized_motif_search(dna, k, t)
        if new_score < score_best:
            best = new_best
            score_best = new_score

    return best


result = repeated_randomized_motif_search(
    "CGCCCCTCTCGGGGGTGTTCAGTAAACGGCCA GGGCGAGGTATGTGTAAGTGCCAAGGTGCCAG TAGTACCGAGACCGAAAGAAGTATACAGGCGT TAGATCAAGTTTCAGGTGCACGTCGGTGAACC AATCCACCAGCTCCACGTGCAATGTTGGCCTA".split(),
    8,
    5,
)
assert (
    score(result) <= score("TCTCGGGG CCAAGGTG TACAGGCG TTCAGGTG TCCACGTG".split()) + 1
), result


if __name__ == "__main__":
    with open("data/rosalind_ba2f.txt", "r") as dataset:
        k, t = [int(d) for d in dataset.readline().rstrip().split()]
        text = [l.rstrip() for l in dataset.readlines()]
    print(*repeated_randomized_motif_search(text, k, t), sep="\n")

File: luiz/src/bin/ba2f.rs
use std::env;
use std::error::Error;
use std::fs;
use std::str;

use rand::distributions::{Distribution, Uniform};

use bioinformatics_algorithms::{profile_matrix, profile_most_probable, score};

pub fn randomized_motif_search(
    text: &[&str],
    k: usize,
    rng: &mut rand::rngs::ThreadRng,
) -> Vec<String> {
    let mut motifs: Vec<String> = text
        .iter()
        .map(|seq| {
            let range = Uniform::from(0..seq.len() - k);
            let rand_pos = range.sample(rng);
            seq[rand_pos..rand_pos + k].into()
        })
        .collect();

    let mut best = motifs.clone(); // TODO avoid this clone // TODO avoid this clone
    let mut score_best = score(&best);

    loop {
        let profile = profile_matrix(&motifs, k);
        motifs = text
            .iter()
            .map(|seq| profile_most_probable(seq, k, &profile))
            .collect();

        let score_motifs = score(&motifs);
        if score_motifs < score_best {
            best = motifs.clone(); // TODO avoid this clone
            score_best = score_motifs
        } else {
            return best.clone(); // TODO avoid this clone
        }
    }
}

pub fn repeated_randomized_motif_search(
    text: &[&str],
    k: usize,
    rng: &mut rand::rngs::ThreadRng,
    times: Option<usize>,
) -> Vec<String> {
    let times = times.unwrap_or(1000);
    let mut best = randomized_motif_search(text, k, rng);
    let mut score_best = score(&best);

    for _ in 1..times {
        let new_best = randomized_motif_search(text, k, rng);
        let new_score = score(&new_best);
        if new_score < score_best {
            best = new_best;
            score_best = new_score;
        }
    }

    best
}

fn main() -> Result<(), Box<dyn Error>> {
    let input: String = env::args()
        .nth(1)
        .unwrap_or("data/rosalind_ba2f.txt".into());
    let data = fs::read_to_string(input)?;
    let mut lines = data.lines();

    let mut ints = lines
        .next()
        .unwrap()
        .split(' ')
        .map(|x| x.parse().unwrap())
        .take(2);
    let k = ints.next().unwrap();
    let _t = ints.next().unwrap();

    let text: Vec<&str> = lines.collect();

    let mut rng = rand::thread_rng();

    for kmer in repeated_randomized_motif_search(&text, k, &mut rng, None) {
        println!("{}", kmer);
    }

    Ok(())
}
